BUG: Fix packbits to correctly handle empty arrays

iwiwi opened the pull request:
This small PR addresses several problems of `numpy.packbits` pointed out in Issue #8324.

As discussed in Issue #8324, `numpy.packbits` have problems with regard to empty input arrays. Specifically:

* the output does not necessarily always have `uint8` dtype,
* the shape of the returned array is inconsistent when `axis` is specified, and
* there is a leak of an object (`new`).

The fundamental reason for these problems is common: function `pack_bits` in `compiled_base.c` handles empty arrays separately, in a rough manner. I could not figure out a good reason why we should handle empty arrays separately. Rather, I found that this part is indeed unnecessary, i.e., without this, `pack_bits` can handle empty arrays correctly, as confirmed in my new tests. So, I propose to remove that part to fix these problems.

iwiwi committed:
BUG: fix packbits and unpackbits to correctly handle empty arrays

juliantaylor commented:
please squash the test and the fix into one commit, it is better to have the testsuite succeed on every commit for easier bisection.

iwiwi commented:
Hi @juliantaylor, thank you for your quick feedback. I squashed the two commits into one commit.

charris commented on numpy/lib/tests/test_packbits.py:
```diff
@@ -17,6 +17,22 @@ def test_packbits():
 
 
+def test_packbits_empty():
+    for dtype in [np.bool, np.uint8, np.int]:
```
Why just these three types? And why just the last dimension zero? Tests should be exhaustive. Admittedly, most aren't but, strangely enough, when improved often reveal errors.

All integer types are available in `np.typecodes['AllInteger']`,  the boolean type is separate. Or you could just use the string `'?bBhHiIlLqQ'` for typecodes. If boolean is acceptable I think the docstring need improvement also.

charris replied:
Also not a good idea to shadow the numpy `dtype` variable. Use something like `dt` instead.

charris replied:
Is it checked anywhere that empty results unpack correctly?

iwiwi commented:
Hi @charris, thank you for the detailed and accurate feedback!

> Why just these three types?
> And why just the last dimension zero?
> Also not a good idea to shadow the numpy dtype variable. Use something like dt instead.
According to your comments, I improved the tests in `test_packbits.py`.

> If boolean is acceptable I think the docstring need improvement also.
I updated the statement in `add_newdoc.py`.

> Is it checked anywhere that empty results unpack correctly?
You are correct --- I added tests for unpack against empty arrays, and indeed, I found a similar bug in unpack. So, I also updated function `unpack_bits` in `compiled_base.c` in the same way as function `pack_bits`.

iwiwi commented:
Sorry for stupid mistake. fixed.

juliantaylor commented:
thanks, very nice pull request.
